### ISSUE TITLE
How to implement graceful handling of source streams' errors?

### ISSUE THREAD

**Issue by Envek** [Faleij/json-stream-stringify#18]
TL;DR: I want to handle errors in source streams, properly terminate JSON for them and also append error info to JSON itself. How can I do this with syntax like this:

```js
new JsonStreamStringify({ result: dataStream, error: wasErrorPromise });
// and to get
// {"result":[{"some":"data"},{"test":"data"}],"error":{"message":"Boom!"}}
// instead of
// {"result":[{"some":"data"},{"test":
```
---

Currently, if source stream emits error, json-stream-stringify also immediately emits error here: https://github.com/Faleij/json-stream-stringify/blob/b5f1b6e95fb88b7f85a1e565539ae36a0abb4ef5/src/JsonStreamStringify.js#L208-L211
In real life (when piping to HTTP response object) it means that connection to client will be reset immediately with partial data sent (and partial means “interrupted inside a string maybe”). I want to gracefully handle it and send a valid JSON to client with partial data and some error message.

So, basically, I want to do something like this:
```js
const { Readable } = require('stream');
const JsonStreamStringify = require('json-stream-stringify');

const dataStream = new Readable({ objectMode: true });
dataStream._read = () => {};

const wasError = new Promise(
  (resolve, reject) => {
    dataStream.on('error', resolve);
    dataStream.on('end', resolve); // will be resolved with undefined eliminating error key from result
  }
);

const jsonStream = new JsonStreamStringify({ result: dataStream, error: wasError });
let result = '';
jsonStream.on('data', (data) => result += data);
jsonStream.on('error', (error) => console.error(error));

dataStream.push({ some: 'data' });
dataStream.push({ more: 'data' });
dataStream.emit('error', { message: 'Boom!' });

console.log(result);
```

Here I want to get following result:
```json
{"result":[{"some":"data"},{"more":"data"}],"error":{"message":"Boom!"}}
```
or, if json-stream-stringify hasn't drained `dataStream` yet (it is totally fine for simplicity):
```json
{"result":[{"some":"data"}],"error":{"message":"Boom!"}}
```
But I get
```json
{"result":[{"some":"data"},{"more":"data"}
```
And in real life, for example when I streaming data from postgres database with [pg-query-stream](https://www.npmjs.com/package/pg-query-stream) and server or connection dies while streaming, data can be truncated between key and value in response (and client will get an error because of connection reset):
```jsonc
{"result":[{"book_ref":"005F0D","book_date":"2017-02-12T00:12:00.000Z"},{"book_ref":
// curl: (18) transfer closed with outstanding read data remaining
```

I wondering of some kind of wrapper I can create for source stream(s) that will re-emit all events except `error` (on which it will just  somehow close itself) and delegate all methods to source stream.

Or may be some kind of such error-handling can be included into json-stream-stringify itself?

Do you have any thoughts, or suggestions? Thanks!

**Comment by Faleij** [comment]
My suggestion is a stream error interceptor, it's a simple passthrough stream that captures any error and gracefully end the stream. The error is available as a promise on the interceptor instance variable ``errorPromise``. For convienience I have also included support for optionally transforming the error before pushing it onto json-stream-stringify, the option is ``transform: (error: Error | any) => any``, see ``new StreamErrorInterceptor(...)`` example below.
```javascript
const { Readable, PassThrough } = require('stream');
const JsonStreamStringify = require('json-stream-stringify');

let i = 0;
const data = [
  { some: 'data1' },
  { some: 'data2' },
  { some: 'data3' },
];
const dataStream = new Readable({
  objectMode: true,
  read() {
    if (i === 2) {
      this.emit('error', { message: 'da error!' });
      // this.destroy({ error: 'da error!' }); // produces same result as above line
      return false;
    }
    i++; // uncomment this line to test StreamErrorInterceptor without errors
    this.push(data.pop() || null);
  },
});

class StreamErrorInterceptor extends PassThrough {
  constructor(options = {}) {
    let { transform, ...opt } = options;
    super({
      objectMode: true, // default to objectMode but does not enforce it
      ...opt, // merge user options
      autoDestroy: false, // force autoDestroy to false to gracefully end stream on error
    });

    let end;
    // create errorPromise property that resolves on error or on end
    this.errorPromise = new Promise((resolve) => {
      end = transform instanceof Function ? (err) => {
        resolve(transform(err));
        setImmediate(() => this.end());
      } : (err) => {
        resolve(err);
        setImmediate(() => this.end());
      };
    });

    // create piping hooks because autoDestroy is set to false
    this.on('pipe', (src) => {
      src.once('error', end);
      src.once('end', end);
    });
    this.on('unpipe', (src) => {
      src.off('error', end);
      src.off('end', end);
    });
  }
}

const errorHandler = new StreamErrorInterceptor({ transform(err) { return err.message; } });
const jsonStream = new JsonStreamStringify({
  result: dataStream.pipe(errorHandler),
  error: errorHandler.errorPromise,
});

let result = '';
jsonStream.on('data', (data) => {
    result += data;
});

jsonStream.on('error', (error) => {
    // will never happen in this example
    console.error('error', error);
});

jsonStream.on('end', () => {
    console.log('result', result);
});
```

**Comment by Envek** [comment]
Thank you very much for complete example!

May be it is a dumb question, but why do we need to postpone ending stream with `setImmediate`? Why it can't be done, well, more immediately after promise resolve?

**Comment by Faleij** [comment]
I am not 100% sure why but I was getting ``write after end`` errors without ``setImmediate``. I think when the error event occurs there might still be data yet to be flushed to the passtrough stream and ending immediately causes a ``write after end`` error.

**Comment by Faleij** [comment]
I have updated the StreamErrorInterceptor class and removed the need for ``setImmediate``. See code comments for explanations.
